### ISSUE TITLE
cannot update row when a column is null

### ISSUE THREAD

**Issue by git4sun** [outworkers/phantom#522]
I got this exception at runtime:com.websudos.phantom.builder.primitives.Primitive$$anon$1 Column xxx is null. Check the source code, it looks like phantom prevents any update on the null columns. But should it be a normal use case since cql allows it? 
My use case is this. I have a table with 12 columns. When creating a record, I only have 3 columns filled with information. Later on I need to add more information to the same record. But whenever I want to update the record with `update.where(_.id eqs id).modify(_.group setTo "g1").... ` I got this exception at runtime.

**Comment by alexflav23** [comment]
HI @git4sun.

Sounds more like you are trying to fetch records instead of updating them, I don't think updates have a chance of creating this problem. I will add tests for your scenario, but you should be using `Optional` columns such as `OptionalIntColumn` and so on if you have things that can be `null`, it's a very big smell to have to deal with `null` in Scala.

Hope this helps, we will investigate on our end as well.

Regards.
Flavian

**Comment by git4sun** [comment]
I described the issue in detail [here](http://stackoverflow.com/questions/38550861/how-to-use-phantom-dsls-optional-columns). I expect the optional columns will handle the use case. But I don't know how.

**Comment by alexflav23** [comment]
Hi @git4sun,

I will add tests for this today in more detail with your use case, sounds like it should work, if it doesn't we will change the relevant underlying update mechanism.

Regards.

**Comment by alexflav23** [comment]
Hi @git4sun,

Tests have been added to reflect this exact scenario, where you update a column with a `null` value, or better said a `None` value. This works without any problems inside phantom. I guess the bit you are not currently doing if there's anything at all I can think of is not using `OptionalColumn` variation for things that can be `null`.

Have a look [at this example](https://github.com/outworkers/phantom/blob/124d27d48b9373b29b3da02ba41b8741e96a3273/phantom-dsl/src/test/scala/com/websudos/phantom/builder/query/db/crud/UpdateTest.scala) and please let me know if you are referring to a different scenario altogether.

Regards.

**Comment by git4sun** [comment]
I think my question is about the default behavior of optional columns. To me, optional means I do not have to specify it. But currently, the optional column is not really "optional". It is either None/null or some value. I need to decide what that value is before updating the database. But if I already know the value, why do I need the optional column? 

 Let's say I have a table User:

`id | name | email | phone`

A user first signed up with name only, I need to create a `User(id, "ally", None, None)`. Next time the user wants to add a phone number. At this point, I don't know the complete record of the user. I only have id and a new phone number. If I create a `User(id, None, None, newphone)`, it will overwrite the name. So I have to read from the table and get a User object. And then set phone number to `newphone`, then update User table with the record. In CQL, I can just write `update user set phone='123456' WHERE id = fb372533-eb95-4bb4-8685-6ef61e994caa`. I know in phantom I can also do it similarly. update.where(_.id eqs u.id).modify(_.phone setTo newphone).future. But since I have 3 optional columns, I have to write 7 update methods to handle different cases. Or I can do then in one update method:

`if(u.name != None) update.where(_.id eqs u.id).modify(_.name setTo u.name).future`

  `if(u.email != None) update.where(_.id eqs u.id).modify(_.email setTo u.email).future`

  `if(u.phone != None) update.where(_.id eqs u.id).modify(_.phone setTo u.phone).future`

But it will be ugly if I want to handle the consistency among these updates. e.g., if the first update fails ..., else if second fails... 
  
I am expecting a optional column behavior that if a value is None, it doesn't do anything to the database. Only if it has some value, the database is updated.

**Comment by alexflav23** [comment]
HI @git4sun,

I finally understood it now. Obviously from a driver perspective we can make no predictions about what the domain looks like and about whether or not when you set `Option` to `None` what you really mean is skip this update, that would be impossible to do at driver level.

We are working to add a `setIfDefined` extra method only available on `OptionalColumn`, which will basically do what you describe, e.g update everything only if the `Option` is a `Some`, otherwise skip the clause entirely, so you don't have to jump through hoops.

It's worth noting most of the time people have separate routes for updating individual parts of their domain model, so you would have one end point for updating the `email` and another for the `phone` and so on, that's why this problem has never been common enough for us to invent a fix, fortunately it looks like what you want is feasible.

Regards,

**Comment by git4sun** [comment]
That sounds good. I understand that it is bad for the driver to make any assumption about the use case. However, would you mind give me a use case where optional columns are used? I mean if I put None to the column, I decided that the value for this column to be null. If I put Some(v), I decided its value to be `v`. What's the difference than using a regular column and use a default value like "null" to represent when I want to set None, and "v" when I want to set Some(v)?

**Comment by alexflav23** [comment]
Hi @git4sun,

This is already fixed in `1.28.2`, you know now use a new method called `setIfDefined` on an option like this:

```scala
update.where(_.id eqs u.id).modify(_.name setIfDefined u.name)
  .and(_.email setIfDefined u.email)
  .and(_.phone setIfDefined u.phone)
  .future
```

As per your use case, this will not generate anything in the query if the `Option` is `None`, so basically the below will just update the name of the user, it will not set other values to `None` or `null` inside Cassandra, instead those parts of the set will not exist in the final query.

```scala
update.where(_.id eqs u.id).modify(_.name setIfDefined Some("test")
  .and(_.email setIfDefined None)
  .and(_.phone setIfDefined None)
  .future
```

**Comment by alexflav23** [comment]
Fixed in 1.28.3, now on GitHub, I'm going to close this and please report back if you have any issues.
